server/sonar-server/src/main/java/org/sonar/server/rule/RegisterRules.java

```diff
@@ -572,4 +576,45 @@ private void update(DbSession session, RuleDefinitionDto rule) {
     dbClient.ruleDao().update(session, rule);
   }
 
+
+  private static void verifyRuleKeyConsistency(RulesDefinition.Context context) {
+    List<RulesDefinition.Rule> allRules = context.repositories().stream()
+      .flatMap(r -> r.rules().stream())
+      .collect(MoreCollectors.toList());
+
+    List<RuleKey> definedRuleKeys = allRules.stream()
+      .map(r -> RuleKey.of(r.repository().key(), r.key()))
+      .collect(Collectors.toList());
```

Comment:
why not a `Set` here?

The only usage of this list is to create a `HashSet` with its content

Reply:
Exact I've added a test on duplicated keys in rule but there is a check already.